### ISSUE TITLE
dim-colours not shown properly

### ISSUE THREAD

**Issue by Atcold** [tmux/tmux#135]
I cannot figure out why, within *Tmux*, `dim`-ed colours do not show properly.
![Issue image](http://i.stack.imgur.com/wFFP5.png)
[Here is the link](http://unix.stackexchange.com/q/234363/38879) to a more thorough description of the problem.

**Comment by nicm** [comment]
Show output of tmux showmsgs -T inside tmux please.

**Comment by Atcold** [comment]
Here's the [gist](https://gist.github.com/Atcold/910f8ce6d90c87b30cc2).

**Comment by nicm** [comment]
Your TERM entry outside tmux doesn't have dim so tmux can't use it, either upgrade terminfo, build your own terminfo entry with dim, or add it with terminal-overrides. The last is probably the easiest:

set -sa terminal-overrides ",*:dim=\E[2m"

**Comment by nicm** [comment]
Probably the application you are using doesn't care (because it works outside without dim), but for some applications you may need to check if the TERM inside tmux supports it, you can see with "infocmp -x|grep dim" inside tmux. If not then you will need to build your own screen or screen-256color terminfo entry (or upgrade terminfo/ncurses). There is an FAQ entry about doing this for italics, doing it for dim is similar.

**Comment by Atcold** [comment]
@nicm, the command you suggested me to run does not work. It simply turned
```
* 35a5032 - (2 days ago) Add install script for Ubuntu Bash - Alfredo Canziani
```
into
```
* 35a5032 - (2 days ago) Add install script for Ubuntu Bash E[2m- Alfredo Canziani
```

**Comment by nicm** [comment]
Looks like you missed off the \ or need to escape it. No dim is always missing, the application outside must ignore terminfo and just send it anyway.

**Comment by Atcold** [comment]
:heart_eyes::heart_eyes::heart_eyes::heart_eyes::heart_eyes:
```bash
set -sa terminal-overrides ",*:dim=\\E[3m"
```
Thank you soo much!!!

**Comment by Atcold** [comment]
@nicm, if you don't mind, do you think you can explain a little further what you're saying?
I am a bit confused about "where" the application is sending the `dim` to. Is it sending it to `DISPLAY`?
What is the role of `TERM` in this?
And how does *Tmux* enter to play in all of this.

**Comment by nicm** [comment]
On most terminals, applications draw on the terminal by sending escape sequences to the terminal, starting with the ASCII ESC character (usually represented as \033 or \e or \E). These are mixed together in with the normal text, so vim might send abc\033[1mdef to show abc then def in bright. For dim, the sequence is almost always \033[2m, but for some things the sequence can differ between different terminals. Terminfo is a database telling the application what it can send to get dim, or bright, or to move the cursor, or other things on the terminal it is running in. TERM is an environment variable which tells the application which entry to look at in the database, so with TERM=xterm if it asks for dim on my computer it will get \033[2m, but on your computer it will get nothing (dim does not exist).

Some applications may bypass terminfo and send \033[2m directly. This mostly works for obvious stuff like dim, bold but it can fail for more complicated stuff. This is what the program you are using is doing, so that is why it works without a dim entry in terminfo.

tmux is both a terminal itself and it runs in a terminal. So tmux needs TERM to be set correctly outside to know what it should send to the outside terminal. Because dim is missing on your system, tmux does not know what to send to get dim text. terminal-overrides tells it to ignore terminfo and just send \033[2m.

Because tmux is itself a terminal, applications inside tmux need to know what to send to tmux to make it draw text in dim. For this they need to look in the screen terminfo entry. So TERM must be "screen" or "screen-256color" inside tmux, or applications will not work correctly. When tmux sees the right escape sequence for dim from an application inside itself, it translates that to dim outside tmux (which it looks up in terminfo using the TERM outside).

tmux can't just pass escape sequences straight through, it has to understand them, because it has to keep a copy of the current terminal state so that it can redraw it if you attach and reattach. So it is important for applications inside tmux to use TERM=screen, or tmux will not necessarily be able to do the right thing.

**Comment by Atcold** [comment]
Wow! Impressive!
Following your line of thought, if on Ubuntu (this time) the `dim

**Comment by nicm** [comment]
The easiest thing is to run "tmux showmsgs -T" inside tmux which shows the terminfo descriptions tmux is using, look for dim:

$ tmux showmsgs -T
Terminal 0: xterm [references=2, flags=0x0]:
   0: AX: (flag) true
...
  25: dim: (string) \033[2m
...

If it says [missing] then it is missing and you will need to use terminal-overrides (at least until Ubuntu gets a newer terminfo).

You can also use infocmp to see the full terminfo if you prefer:

$ infocmp -x xterm|grep dim
        cvvis=\E[?12;25h, dch=\E[%p1%dP, dch1=\E[P, dim=\E[2m,

**Comment by Atcold** [comment]
Hmm... No luck :disappointed:
```bash
$ tmux showmsgs -T
tmux: unknown option -- T
usage: show-messages [-t target-client]
```
and
```bash
$ infocmp -x xterm|grep dim
# no output
```

**Comment by nicm** [comment]
Your tmux is too old for showmsgs -T, use tmux info instead.

In any case, infocmp shows your xterm entry doesn't have dim, so you will need to use terminal-overrides to add it.

**Comment by Atcold** [comment]
OK, *Tmux* reported that
```
25: dim: [missing]
```
I'm already overwriting the `dim` directive, like I do on my Mac, but no luck.
`dim` is not shown within *Tmux* or outside it.

I currently am on *GNOME Terminal 3.6.2*, and [here](http://misc.flogisoft.com/bash/tip_colors_and_formatting#terminals_compatibility) it is said that `dim` is supported. (Sorry for bugging you with *Tmux* unrelated questions...)

**Comment by nicm** [comment]
What tmux version doesn't work? What have you got in .tmux.conf?

**Comment by nicm** [comment]
It probably isn't the GNOME Terminal version that matters but the ncurses or terminfo version.

**Comment by ELLIOTTCABLE** [comment]
I'm having the same issue; and despite all of the thorough explanation above, I can't figure out *a solution*.

I have the latest OS X (as of the writing, El Capitan), and wish to use Terminal.app. *How*, precisely, do I either:

1. ‘upgrade terminfo’,
2. build my own terminfo for Terminal.app,
3. or ‘add it with `terminal-overrides`?

I've tried executing `set -sa terminal-overrides ",*:dim=\E[2m" ; tmux`, and `echo -e '\e[2atest'` still doesn't work; I've also tried running the same command *inside* a tmux shell, and again, it's not displayed.

I suspect this deserves a writeup in the README; I'm probably not the only `tmux` user using Terminal.app 😜
